language/move-prover/tests/sources/functional/macro_verification.move

```diff
@@ -0,0 +1,100 @@
+// flag: --trace
+
+// CVC5 does currently not terminate on this:
+// exclude_for: cvc5
+
+/// This file contains some simulated expansions of functional macros and sketches how they can be verified.
+module 0x42::FunMacros {
+    use Std::Vector;
+
+    /// Simulates `foreach!(v, |x| *x = *x + 1 spec  { ensures x == old(x) + 1; })`
+    /// where `macro foreach!<T>(v: &mut vector<T>, action: |&mut T|)`
+    ///
+    /// For the foreach macro the loop invariant can be synthesized from a post condition associated
+    /// with the lambda.
+    fun foreach(v: &mut vector<u64>) {
+        let i = 0;
+        while (i < Vector::length(v)) {
+            let x = Vector::borrow_mut(v, i);
+            *x = *x + 1;
+            i = i + 1;
+        }
+        spec {
+            invariant i >= 0 && i <= len(v);
+            invariant len(v) == len(old(v));
+            invariant forall j in 0..i: v[j] == old(v)[j] + 1; // lambda substituted
+            invariant forall j in i..len(v): v[j] == old(v)[j];
+        };
+    }
+    spec foreach {
+        ensures len(v) == len(old(v));
+        ensures forall i in range(v): v[i] == old(v)[i] + 1; // succeeds
+        ensures forall i in range(v): v[i] == old(v)[i] + 2; // fails
+    }
+
+    /// Simulates `reduce!(v, 0, |sum, x| *sum = *sum + *x ) spec { invariant sum == old(sum) + x; })`
```

Comment:
Note that the guarantees of the borrow checker also pop up here.  The invariant "sum == old(sum) + x" is valid only because while sum is a mutable reference x is immutable.  Therefore, the borrow checker guarantees that a modification via sum leaves x unchanged.